Expand development procedures (#74)
yt-project/yt-4.0-paper: content/15.development_procedure.md

One increasing issue with the code review process is ensuring that changes are reviewed with appropriate urgency; larger pull requests tend to languish without review, as the requirements for review necessarily add burden to the maintainers.
"Bugfix" changes formally require only one reviewer, whereas the `yt` guidelines suggest that larger changes require review from two different team members.

> Recently, we have begun utilizing the ["auto-merge" feature](https://github.blog/changelog/2021-02-04-pull-request-auto-merge-is-now-generally-available/) provided by GitHub.

Reviewer: I think I would reorganise this paragraph to tell the problem first, and then show what solution we adopted. I find it would engage the reader more. Arguably, the problem is also more fundamental than its (current) solution.

One of the most pressing bottlenecks with code review is that the time it takes the continuous integration tests to pass is much longer than the typical time span during with code review takes place.
This results in a pull request being merged, and then whatever pull request is next to be reviewed has to wait until the tests (now updated with the newly accepted pull request) pass.
By queuing up pull requests for merging, it allows maintainers to mark a set of pull requests as ready to be merged, and then when they meet the appropriate (automated and asynchronous) criteria, they will be merged.

### Code Styling and Linting

For code included in ``yt``, a set of styles are enforced.
We use the automated code styling tool [``black``](https://black.readthedocs.io/en/stable/), which has been designed to maximize readability with as few free parameters as possible.
(In many ways, the fact that most of the ``yt`` code developers did not utilize this style before it was enforced likely enabled its uptake, as it was seen as an choice that "made everyone compromise.")

> In addition to code formatting, other automated methods for code consistency are applied; these include ensuring that anywhere a Python "f-string" could be used it is, that all ``import`` statements are sorted (according to alphabetical order within a first/second/third-party categorization).

Reviewer: the part on f-strings is a bit misleading. We do not use flynt anymore, but we rely on pyupgrade for string formatting among other things. Pyupgrade is however intentionally "shy" on fstring senforcement and it only uses them where it actually makes code more concise (see https://github.com/asottile/pyupgrade#f-strings)

Author: How's the new version?

For the purposes of checking for common code mistakes, we apply [``flake8``](https://flake8.pycqa.org/en/latest/), which reports unused variables, undefined variables, and other style violations that cannot be automatically corrected.

All changes that can be applied via automation (specifically, code formatting) are accessible from within the GitHub pull request interface.
This allows "drive-by" contributions to have their pull requests updated inline by an automated process, reducing the need to manually install packages to apply the changes.

> 

Reviewer: I think if we're going to mention Github's auto-merge, that we've been using for less than a week now, pre-commit deserves a citation too. Without it, syncing black + isort + pyupgrade across developers and github would be a nightmare, and we probably would have given up *or* lost tons of time managing the tool belt.

Author: Made #75

### Type Hinting

Code included in ``yt`` is not required to utilize [type hinting](https://docs.python.org/3/library/typing.html).

> However, new contributions are allowed to include type hinting, and work is underway to develop an ontology of types as well as a base set of utilities and types that can be used both internally and by external package developers.

Reviewer: I don't think this work is mature enough that we can even start advertising it as "usable by external package developers". I think we *need* to keep it internal for now because it's bound to be experimental (typing in Python is the wild west atm), and it would be a pity to start breaking downstream or settling for (bad) early solutions just because we invited other people to use them.

Reviewer: sure. How about this ?
```suggestion
However, new contributions are allowed to include type hinting, and work is underway to develop an ontology of types as well as a base set of utilities and types that would be used both internally, and possibly be made available to external package developers.
```

One increasing issue with the code review process is ensuring that changes are reviewed with appropriate urgency; larger pull requests tend to languish without review, as the requirements for review necessarily add burden to the maintainers.
"Bugfix" changes formally require only one reviewer, whereas the `yt` guidelines suggest that larger changes require review from two different team members.

> One of the most pressing bottlenecks with code review is that the time it takes for tests to pass is much longer than the typical time span during with code review takes place.

Reviewer: Looks like a typo to me ?
```suggestion
One of the most pressing bottlenecks with code review is that the time it takes for tests to pass is much longer than the typical time span during which code review takes place.
```

> This results in a pull request being merged, and then whatever pull request is next to be reviewed has to wait until the tests (now updated with the newly accepted pull request) pass.

Reviewer: This sentence doesn't resonate with my personal experience as a reviewer; I don't think that merging a PR has any noticeable impact on reviewing other PRs in the following minutes because they will typically have run their jobs already. The actual limitation can be seen in a different situation, I think, where reviewers and authors have to iterate over a PR, as the CI can't go over more than a couple cycles per hour.

Author: Hmm, usually when I'm reviewing a PR I want all the tests to remain green, so if I merge one and it causes tests to rerun, we have to wait.
